src/storage-controller/src/lib.rs

```diff
@@ -933,46 +929,80 @@ where
+                            let last_status_per_id = self
+                                .partially_truncate_status_history(
+                                    IntrospectionType::SourceStatusHistory,
+                                )
+                                .await;
+
+                            let status_col = healthcheck::MZ_SOURCE_STATUS_HISTORY_DESC
+                                .get_by_name(&ColumnName::from("status"))
+                                .expect("schema has not changed")
+                                .0;
+
+                            self.collection_status_manager.extend_previous_statuses(
+                                last_status_per_id.into_iter().flatten().map(|(id, row)| {
+                                    (
+                                        id,
+                                        row.iter()
+                                            .nth(status_col)
+                                            .expect("schema has not changed")
+                                            .unwrap_str()
+                                            .into(),
+                                    )
+                                }),
                             )
-                            .await;
                         }
                         IntrospectionType::SinkStatusHistory => {
-                            self.partially_truncate_status_history(
-                                IntrospectionType::SinkStatusHistory,
+                            let last_status_per_id = self
+                                .partially_truncate_status_history(
+                                    IntrospectionType::SinkStatusHistory,
+                                )
+                                .await;
+
+                            let status_col = healthcheck::MZ_SINK_STATUS_HISTORY_DESC
+                                .get_by_name(&ColumnName::from("status"))
+                                .expect("schema has not changed")
+                                .0;
+
+                            self.collection_status_manager.extend_previous_statuses(
+                                last_status_per_id.into_iter().flatten().map(|(id, row)| {
+                                    (
+                                        id,
+                                        row.iter()
+                                            .nth(status_col)
+                                            .expect("schema has not changed")
+                                            .unwrap_str()
+                                            .into(),
+                                    )
+                                }),
                             )
-                            .await;
                         }
                         IntrospectionType::PrivatelinkConnectionStatusHistory => {
-                            // Truncate the private link connection status history table and
```

Comment:
I think it'd be helpful to leave a comment like this somewhere, explaining that for the privatelink statuses we use the `occured_at` timestamp as the comparison for whether to store new events, but for other status tables we use the `status` column value itself.